packages/runtime/container-runtime/src/summarizerHeuristics.ts

```diff
@@ -39,46 +59,70 @@ export class SummarizeHeuristicData implements ISummarizeHeuristicData {
+
+        this.numSystemOpsBefore = this.numSystemOps;
+        this.numNonSystemOpsBefore = this.numNonSystemOps;
     }
 
     public markLastAttemptAsSuccessful() {
         this._lastSuccessfulSummary = { ...this.lastAttempt };
+
+        this.numSystemOps -= this.numSystemOpsBefore;
+        this.numSystemOpsBefore = 0;
+
+        this.numNonSystemOps -= this.numNonSystemOpsBefore;
+        this.numNonSystemOpsBefore = 0;
     }
 }
 
 /**
  * This class contains the heuristics for when to summarize.
  */
 export class SummarizeHeuristicRunner implements ISummarizeHeuristicRunner {
-    private readonly idleTimer: Timer;
+    private readonly idleTimer: Timer | undefined;
+    private readonly runSummarize: (reason: SummarizeReason) => void;
 
     public constructor(
         private readonly heuristicData: ISummarizeHeuristicData,
         private readonly configuration: ISummaryConfiguration,
-        private readonly trySummarize: (reason: SummarizeReason) => void,
+        trySummarize: (reason: SummarizeReason) => void,
         private readonly minOpsForAttemptOnClose = 50,
+        private readonly summarizeStrategies: ISummarizeHeuristicStrategy[] = getDefaultSummarizeHeuristicStrategies(),
+        useIdleTimerStrategy: boolean = true,
     ) {
-        this.idleTimer = new Timer(
-            this.configuration.idleTime,
-            () => this.trySummarize("idle"));
+        if (useIdleTimerStrategy) {
+            this.idleTimer = new Timer(
+                this.configuration.idleTime,
+                () => this.runSummarize("idle"));
+        }
+
+        this.runSummarize = (reason: SummarizeReason) => {
+            this.idleTimer?.clear();
+
+            // We shouldn't attempt a summary if there are no new processed ops
+            const opsSinceLastAck = this.opsSinceLastAck;
+            if (opsSinceLastAck > 0) {
+                trySummarize(reason);
+            }
+        };
     }
 
     public get opsSinceLastAck(): number {
         return this.heuristicData.lastOpSequenceNumber - this.heuristicData.lastSuccessfulSummary.refSequenceNumber;
     }
 
+    public start() {
+        this.idleTimer?.start();
+    }
+
     public run() {
-        const timeSinceLastSummary = Date.now() - this.heuristicData.lastSuccessfulSummary.summaryTime;
-        const opsSinceLastAck = this.opsSinceLastAck;
-        if (timeSinceLastSummary > this.configuration.maxTime) {
-            this.idleTimer.clear();
-            this.trySummarize("maxTime");
-        } else if (opsSinceLastAck > this.configuration.maxOps) {
-            this.idleTimer.clear();
-            this.trySummarize("maxOps");
-        } else {
-            this.idleTimer.restart();
+        for (const strategy of this.summarizeStrategies) {
+            if (strategy.shouldRunSummarize(this.configuration, this.heuristicData)) {
+                return this.runSummarize(strategy.summarizeReason);
+            }
         }
+
+        this.idleTimer?.restart();
```

Comment:
I mentioned it somewhere before, but I do not think it's addressed. The code before would not call run() method for summarize and summarize ack ops. This would mean we would not be summarizing these ops in a loop infinitely.

Now, given that we call run() for all system ops, this area is regressed. I'd love to see a UT that catches that case, and a fix. Fix can be either not call run for those two ops from handleOp() handlers. Maybe there are other ways to avoid it.